db/view/view.cc

```diff
@@ -1746,26 +1746,33 @@ future<> view_update_generator::mutate_MV(
             stats.view_updates_pushed_remote += updates_pushed_remote;
             cf_stats.total_view_updates_pushed_remote += updates_pushed_remote;
             schema_ptr s = mut.s;
-            future<> view_update = apply_to_remote_endpoints(_proxy.local(), std::move(view_ermp), *target_endpoint, std::move(remote_endpoints), std::move(mut), base_token, view_token, allow_hints, tr_state).then_wrapped(
-                    [s = std::move(s), &stats, &cf_stats, tr_state, base_token, view_token, target_endpoint, updates_pushed_remote,
-                            units = sem_units.split(sem_units.count()), apply_update_synchronously] (future<>&& f) mutable {
-                if (f.failed()) {
-                    stats.view_updates_failed_remote += updates_pushed_remote;
-                    cf_stats.total_view_updates_failed_remote += updates_pushed_remote;
-                    auto ep = f.get_exception();
-                    tracing::trace(tr_state, "Failed to apply view update for {} and {} remote endpoints",
+            // The "delay_before_remote_view_update" injection point can be
+            // used to add a short delay (currently 0.5 seconds) before a base
+            // replica sends its update to the remove view replica.
```

Comment:
Isn't using a delay asking for a failure on a busy system? In the case of tests, probably false negatives.

Probably, such a fault injection point should pause to coordinate on some named http endpoint. For example:

1. test setup registers an http endpoint (provided by the test script) with the API server
2. inject() calls the http endpoint with the input string
3. the http server does something and returns a response
4. based on the response, the injector continues normally or throws an exception

This allows the test system to inject any action, not just a delay.

We won't do this now, but maybe we'll find the time to reorganize this one day.

Reply:
You're absolutely right - a fixed delay *may* add false negatives - if the test machine is so slow that it can't perform a single read in 0.5 seconds, it may miss a bug. But if there is a bug, in some of the runs it won't be missed, and that's enough. Moreover, during the **development** process - when I wanted to reproduced the bug - I ran it on an idle machine.

Please note that this test really needs a finite delay - and not a very long delay or an infinite pause:
1. When synchronous_update=false (the bug), the delay delays the view updates, and the immediate read doesn't see the data.
2. When synchronous_update=true (the fix), the delay delays the view updates, but the write waits for it, and when the write returns, it sees the data.

Maybe there's a better way to do it, but I didn't think of a different approach that is reasonable. I also tried to work within what the injection framework (that is the first time I ever used...) gives me and not to reinvent it.